### ISSUE TITLE
Redundant bracket warning suggests broken code

### ISSUE THREAD

**Issue by boblehest** [ndmitchell/hlint#909]
```
Redundant bracket
Found:
f = case 0 of
      _ | n <- (0 :: Int) -> n
Why not:
f = case 0 of
      _ | n <- 0 :: Int -> n
```
The suggested alternative doesn't parse.

**Comment by ndmitchell** [comment]
Thanks for the report. Agreed, this is a bug - just looking into where else the same bug might be present (figuring out brackets in the presence of pattern guards and view patterns is tricky.

**Comment by ndmitchell** [comment]
Thanks! I've just pushed some fixes. Basic issue is that `(a :: b) -> c` turns into a type if you remove the brackets. Anywhere `->` is expected (RHS of a case or RHS of a view pattern) you can run into problems. I've dialed back the strength of the warning, almost certainly missing some valid warnings, to make sure we don't get it wrong. Can dial it up in future now we have these test cases added.

**Comment by ndmitchell** [comment]
While this code is in HEAD, and I'd usually make a release for incorrect warnings, at the moment we are in a transitional state, so it will probably be another week before the release comes out.
